### ISSUE TITLE
Training with DP?

### ISSUE THREAD

**Issue by DENVERCODER999** [Team-TUD/CTAB-GAN-Plus#2]
Great work! I don't seem to be able to find training with DP to be possible with this codebase. Am I overlooking something, or is this forthcoming?

**Comment by zhao-zilong** [comment]
@DENVERCODER999  DP will come soon! Current CTAB-GAN-Plus code is CTAB-GAN + (WGAN+GP loss) + general transform encoding. The training is more stable.

**Comment by DENVERCODER999** [comment]
Second follow-up question, more conceptual. I see that in the paper and code that you have additional losses and training mechanisms (information loss, training-by-sampling) that, at first glance, makes the privacy analysis a bit more involved. For example, imagine choosing a conditional vector for which there is only a single valid batch from the dataset, intuitively it seems like the amplification via subsampling is nonexistent since $m = n$. Or consider the information loss, where it seems mathematically impossible to compute a per-sample gradient, since different chosen examples will interact with each other.

Is this all dealt with? I don't see any additional privacy analysis in the paper.

**Comment by rodrigo-castellon** [comment]
Hi,
I see, thanks. I'm still a bit confused.

> the additional losses (generator loss, information loss, classification loss) are added on the training of generator. Since our DP is added only on discriminator, it is not more complicated.

In the traditional GAN training setup, you only train the discriminator with DP-SGD because, since post-processing applies, any further update you perform on the generator that depends only on the weights of the discriminator (and importantly not on real examples) does not need to be clipped and noised. But in this setup, the information loss and classification loss don't take advantage of the post-processing property of the discriminator weights and update the generator weights directly based on real examples, so I think training with some sort of clip+noise setup is still necessary, correct me if I'm wrong.

> And also, we didn't consider the per-sample gradient. But I think I got what is your concern. Notice that DP is not just adding noise, it also clips gradients. Therefore one particular sample did not reveal more information than other data.

I see. What I meant here is that when you do standard DP-SGD, we try to stick to computation graphs for which the loss is computed as a mean over individual losses in a batch, and where the loss for an individual example is well-defined and independent of other examples. However, when you do a computation like (num_examples, feature) -> mean -> (feature,) -> ... as done in the information loss, I think this requires an additional privacy mechanism, which is why we usually avoid BatchNorm in favor of comparable GroupNorm, etc. when doing DP-SGD. Otherwise, a single example could have an unbounded impact on the update to the generator's parameters, which could allow you to figure out whether an example was present in the dataset or not. Again, correct me if I'm wrong.

**Comment by zhao-zilong** [comment]
Hi @rodrigo-castellon  
Thanks for the message! I think your comment on the information loss part makes sense, I will talk about that with our team and may give you a more detailed response later.
